test-app/tests/integration/components/headless-form-native-validation-test.gts

```diff
@@ -0,0 +1,754 @@
+      const invalidHandler = sinon.spy();
+
+      await render(<template>
+        <HeadlessForm @data={{data}} @onInvalid={{invalidHandler}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await click('[data-test-submit]');
+
+      assert.true(
+        invalidHandler.calledOnce,
+        '@onInvalid was called when required field in empty'
+      );
+    });
+
+    test('onSubmit is called when validation passes', async function (assert) {
+      const data: TestFormData = {};
+      const submitHandler = sinon.spy();
+
+      await render(<template>
+        <HeadlessForm @data={{data}} @onSubmit={{submitHandler}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+          </form.field>
+          <form.field @name="lastName" as |field|>
+            <field.label>Last Name</field.label>
+            <field.input required data-test-last-name />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await fillIn('input[data-test-first-name]', 'Nicole');
+      await fillIn('input[data-test-last-name]', 'Chung');
+      await click('[data-test-submit]');
+
+      assert.true(submitHandler.called, '@onSubmit has been called');
+    });
+
+    test('validation errors are revalidated on submit', async function (assert) {
+      const data: TestFormData = {};
+
+      await render(<template>
+        <HeadlessForm @data={{data}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+            <field.errors data-test-first-name-errors />
+          </form.field>
+          <form.field @name="lastName" as |field|>
+            <field.label>Last Name</field.label>
+            <field.input data-test-last-name />
+            <field.errors data-test-last-name-errors />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      assert
+        .dom('[data-test-first-name-errors]')
+        .doesNotExist(
+          'validation errors are not rendered before validation happens'
+        );
+      assert
+        .dom('[data-test-last-name-errors]')
+        .doesNotExist(
+          'validation errors are not rendered before validation happens'
+        );
+
+      await click('[data-test-submit]');
+
+      assert
+        .dom('[data-test-first-name-errors]')
+        .exists({ count: 1 }, 'validation errors appear when validation fails');
```

Comment:
I like that this test is in here [based on reading this comment](https://github.com/CrowdStrike/ember-headless-form/pull/29/files/a253d4230b82f6097c10e4299d7013e42351a3fe#r1097548956)